DEP: create new package.json and composer. json with approved packages.

lukastocker opened the pull request:
The list about the approved dependency packages can be found here: [JourFixe-2024-01-22](https://docu.ilias.de/ilias.php?baseClass=ilwikihandlergui&cmdNode=16n:r6&cmdClass=ilobjwikigui&cmd=viewPage&ref_id=1357&page=JourFixe-2024-01-22#ilPageTocA2513)

As mentioned in [PR 6653](https://github.com/ILIAS-eLearning/ILIAS/pull/6653#issuecomment-1964466578) I have not included the simplesamlphp package as it's currently not working with the Symfony Console package version 7. 

As you can see a commit with a new phpunit.xml file was also created, as it would not have worked otherwise. Please also see the change log of PHPUnit 10+, as many functions are deprecated: 
- [Change log](https://github.com/sebastianbergmann/phpunit/blob/10.5/ChangeLog-10.5.md) 
- [Deprecrations](https://github.com/sebastianbergmann/phpunit/blob/10.5/DEPRECATIONS.md)

kergomard committed:
Kergomard: Add sabre/dav As Dependency

Usage:
```components/ILIAS/WebDAV

Reasoning:
```Sabre/DAV``` is used in big projects e.g. in Nextcloud.

Maintenance:
- Release 4.5 released in 2023, approximately one minor
and one major release per year.
- Irregular, but continuous commit activity be multiple contributors
(see: https://github.com/sabre-io/dav/commits/master)
- Not much information on Security.

Links:
Packagist: https://packagist.org/packages/sabre/dav
GitHub: https://github.com/sabre-io/dav
Documentation: https://sabre.io/dav/

mjansenDatabay committed:
Composer: Add `phpmailer/phpmailer` as dependency

Composer: Add `phpmailer/phpmailer` as dependency

This PR adds `phpmailer/phpmailer` as composer dependency.

Usage:
* `components/ILIAS/Mail`

Wrapped By:
* `\ilMailMimeTransportBase`

Reasoning:
* `PHPMailer` is a de facto standard for sending emails with PHP. It supports multiple transport channels as SMTP, mail or sendmail. Furthermore, it supports multiple authentication methods and provides proper APIs for attachment handling (and multipart handling in general), encoding and encryption.
* We should not/do not want to reinvent the wheel in ILIAS and should rely on this library crafted by these email standard specialists.

Maintenance:
* `PHPMailer` is actively maintained by multiple contributors. There is recent activity (even today, see: https://github.com/PHPMailer/PHPMailer/commits/master).
* Security issues are always fixed in a timely manner followed by new releases.

Links:
* Packagist: https://packagist.org/packages/phpmailer/phpmailer
* GitHub: https://github.com/PHPMailer/PHPMailer
* Documentation: https://github.com/PHPMailer/PHPMailer/wiki / https://github.com/PHPMailer/PHPMailer/tree/master/docs / https://phpmailer.github.io/PHPMailer/

mjansenDatabay committed:
Composer: Add `simplesamlphp/simplesamlphp` as dependency

This PR adds `simplesamlphp/simplesamlphp` as composer dependency.

Usage:
* `components/ILIAS/Saml`

Wrapped By:
* `\ilSimpleSAMLphpWrapper` (implements `\ilSamlAuth`)
* `\ilSimpleSAMLphplIdpDiscovery` (implements `\ilSamlIdpDiscovery`)

Reasoning:
* `SimpleSAMLphp` is an award-winning application written in native PHP that deals with secure authentication. The main focus of `SimpleSAMLphp` is providing support for SAML 2.0 as both "Service Provider" and "Identity Provider".
* SAML is a very complex standard (in my personal opinion, see https://datatracker.ietf.org/doc/html/rfc7522 and linked documents) and we should not try to/do not want to reinvent the wheel in ILIAS. So we should rely on this library crafted by these authentication/security standard specialists.

Maintenance:
* `SimpleSAMLphp` is actively maintained by multiple contributors (see: https://simplesamlphp.org/contrib/). There is recent activity (even this week, see: https://github.com/simplesamlphp/simplesamlphp/commits/master).
* Security issues are always fixed in a timely manner followed by new releases. A proper security process is implemented: https://simplesamlphp.org/security/ .

Outlook:
* We are currently working on upgrading the library to the new major release (2.x) to receive all future (security) bugfixes and to be compliant with future PHP versions.

Links:
* Packagist: https://packagist.org/packages/simplesamlphp/simplesamlphp
* GitHub: https://github.com/simplesamlphp/simplesamlphp
* Documentation: https://simplesamlphp.org/docs/

mjansenDatabay committed:
Composer: Add `ezyang/htmlpurifier` as dependency

This PR adds `ezyang/htmlpurifier` as composer dependency.

Usage:
* `components/ILIAS/Html`

Wrapped By:
* `\ilHtmlPurifierAbstractLibWrapper` (implements `\ilHtmlPurifierInterface`)

Reasoning:
* `HTMLPurifier` is a standards-compliant HTML filter library written in PHP. It will not only remove all malicious code (better known as XSS) with a thoroughly audited, secure yet permissive whitelist, it will also make sure your documents are standards compliant, something only achievable with a comprehensive knowledge of W3C's specifications.
* We should not try to/do not want to reinvent the wheel in ILIAS and rely on this security library when filtering HTML code.

Maintenance:
* `HTMLPurifier` is actively maintained by multiple contributors. There is recent activity (even last week, see: https://github.com/ezyang/htmlpurifier/commits/master).
* Security issues are always fixed in a timely manner followed by new releases. Similar to ILIAS the project provides a mailing list for security advisories: http://htmlpurifier.org/contact .

Links:
* Packagist: https://packagist.org/packages/ezyang/htmlpurifier
* GitHub: https://github.com/ezyang/htmlpurifier
* Documentation: http://htmlpurifier.org/docs

mjansenDatabay committed:
Composer: Add `cweagans/composer-patches` as dependency

This PR adds `cweagans/composer-patches` as composer dependency.

Usage:
* During `composer install` and `composer upgrade`

Wrapped By:
* Not applicable

Reasoning:
* `Composer Patches` is a composer plugin which allows us to apply patches to our libraries.
* ILIAS still requires a few patches to be applied after the libraries have been fetched, see: vendor/composer/patches

Maintenance:
* `Composer Patches` is still maintained, mostly by "cweagans (Cameron Eagans)" and supported by other contributors. There is not much activity in general, but the plugin is so to speak "feature-complete". There is activity if there are security vulnerabilities (`Dependabot` pull requests are merged) and if changes were made in `Composer` itself which require the plugin to adapt.
* The risk of relying on this library is not big. It is only used after the installation/upgrade of our composer libraries. If we could not use this library anymore (because the project is not active anymore of if there are major security issues), we will find another way to apply the patches.

Links:
* Packagist: https://packagist.org/packages/cweagans/composer-patches
* GitHub: https://github.com/cweagans/composer-patches
* Documentation: https://docs.cweagans.net/composer-patches/

mjansenDatabay committed:
Composer: Add `captainhook/captainhook` and `captainhook/plugin-composer` as dependency

This PR adds `captainhook/captainhook` and `captainhook/plugin-composer` as composer dependencies.

Usage:
* Executed during multiple `Composer` and `Git` actions (depending on our shared and possible individual configuration files)

Wrapped By:
* Not applicable

Reasoning:
* `CaptainHook` is an easy to use and very flexible `Git hook` library for PHP developers. It enables us to configure our `Git hook` actions in a simple (and shared) JSON file.
* `CaptainHook` helps us to enforce PHP code style and language file structure policies. Generally speaking it allows the execution of any script for arbitrary actions, so developers can easily extend the shared configuration by e.g. executing unit tests before a `git push`, or checking commit messages against a configured ruleset.

Maintenance:
* `CaptainHook` is well maintained (see: https://github.com/captainhookphp/captainhook/commits/main) and widely used in the PHP community.
* The risk of relying on this library is small. It is a development dependency all mandatory code policies are enforced again by our GitHub CI pipeline on the Git remote server.

Links:
* Packagist: https://packagist.org/packages/captainhook/captainhook / https://packagist.org/packages/captainhook/plugin-composer
* GitHub: https://github.com/captainhookphp/captainhook / https://github.com/captainhookphp/plugin-composer
* Documentation: https://captainhookphp.github.io/captainhook/

mjansenDatabay committed:
Composer: Add `mikey179/vfsstream` as dependency

This PR adds `mikey179/vfsstream` as composer dependency.

Usage:
* Used when executing the ILIAS unit test suite.

Wrapped By:
* Not applicable

Reasoning:
* `vfsStream` is a stream wrapper for a virtual file system that may be helpful in unit tests to mock the real file system.

Maintenance:
* There is not much development activity (see: https://github.com/bovigo/vfsStream/graphs/commit-activity), so it might occur that there will be issues with upcoming PHP versions.
* The risk of relying on this library is small. It is a development dependency and only a small number of unit tests rely on a mocked file system. If more and more components in ILIAS use the `IRSS`, the usage of PHP filesystem functions will decrease accordingly, so the corresponding unit tests might get obsolete.

Links:
* Packagist: https://packagist.org/packages/mikey179/vfsstream
* GitHub: https://github.com/bovigo/vfsStream
* Documentation: https://github.com/bovigo/vfsStream/wiki

mjansenDatabay committed:
Composer: Add `phpstan/phpstan` as dependency

This PR adds `phpstan/phpstan` as composer dependency.

Usage:
* `PHPStan` finds issues in the ILIAS code without writing tests. It detects architectural smells in our application based on rule definitions (see: scripts/PHPStan/Rules). It is used in the GitHub CI pipeline (see: .github/workflows/legacy-ui.yml). `PHPStan` makes use of the `AST` (abstract syntax tree) and is capable to analyze our ILIAS PHP code statically.

Wrapped By:
* Not applicable

Reasoning:
* `PHPStan` is one of the two or three industry standards (beside `psaml` and `phan`) to statically analyze PHP code. In ILIAS it is/can be used as a PHP alternative to `dicto`. `PHPStan` is also implicitly used by other libraries (like `rector`). Providing a toolset to enable rule based static analyses is a hard task (in time/costs for the maintenance), so we should not try to reinvent the wheel here by trying to do this analyses ourselves.

Maintenance:
* `PHPStan` is a very active project (see: https://github.com/phpstan/phpstan/graphs/commit-activity), code was committed even this week (see: https://github.com/phpstan/phpstan/commits/1.11.x). There is one main contributor (ondrejmirtes) supported by other open source enthusiasts (and financially by some well known organizations).
* The risk of relying on this library is relatively small. It is a development dependency and our production code does not rely on this library. But: The project mainly relies on one contributor and there might be funding issues in future (a commercial plan has been indroduced, see: https://phpstan.org/blog/introducing-phpstan-pro).

Outlook:
* The existing UI rules could be further extended by  https://github.com/mjansenDatabay/ILIAS/commit/5e4dfeaf4a471de51712681f078758ed9f44b26c / https://github.com/mjansenDatabay/ILIAS/commit/a0907099d64080d9c0959436304d2b5a536866cd

Links:
* Packagist: https://packagist.org/packages/phpstan/phpstan
* GitHub: https://github.com/phpstan/phpstan
* Documentation: https://phpstan.org/user-guide/getting-started

chfsx committed:
Composer: Add `ramsey/uuid` as dependency

klees committed:
Composer: Add filp/whoops  as dependency

This PR adds `filp/whoops ` as composer dependency.

Usage:
* Used for formatting error messages for consumation by developers via browser
  and for error reports.

Wrapped By:
* Not applicable. This is used in error handling code that does not provide
  interfaces to other ILIAS code.

Reasoning:
* `filp/whoops` is a mature and stable library that produces comprehensible
  output of errors, exceptions and the environment they have been created in.
  Good error reporting boosts productivity of developers and admins alike.

Maintenance:
* The library and its are stable for a long time now. Major version hasn't been
  bumped for > 7 years now.
* Patch version is bumped regularly to deliver fixes. Last version has been published
  2023-11-03. Hence maintenance seems to be stable.
* There is a corporate sponsor of this library.

Links:
* Packagist: https://packagist.org/packages/filp/whoops
* GitHub: https://github.com/filp/whoops

klees committed:
Composer: Add pimple/pimple as dependency

This PR adds `pimple/pimple` as composer dependency.

Usage:
* Provides the dependency injection container that is used as a service locator
  throughout the system.
* Will most likely be used as the DIC solution for the new initialisation scheme
  introduced with the [component revision](docs/development/components-and-directories.md).

Wrapped By:
* ILIAS\DI\Container

Reasoning:
* Pimple provides a light weight and simple dependency injection container, which
  is yet quite powerful. The interface is modelled along \ArrayAccess, hence it
  can be easily mocked for testing.
* Dependency injection is ILIAS' future way to express dependencies between
  services and components.

Maintenance:
* The library has been stable for a long time and didn't receive updates since 2021.
* The actual code is only about 150 LOC and hence could be easily maintained by us,
  if necessary.

Links:
* Packagist: https://packagist.org/packages/pimple/pimple
* GitHub: https://github.com/silexphp/Pimple

klees committed:
Composer: Add symfony/console as dependency

This PR adds `symfony/console` as composer dependency.

Usage:
* Provides the TUI for the setup console application.

Wrapped By:
* Not applicable, functionality is only used internally in the setup and not
  provided to other ILIAS components.

Reasoning:
* The library can provide nice interfaces for console applications. The setup is
  the first thing administrators will encounter when installing ILIAS, hence we
  should try to look and feel good there to not put people off.
* The library only provides plumbing to access actual setup functionality, the
  setup only depends on the library for its user interface. Usage of the library
  is encapsulate tightly.

Maintenance:
* The library is under active development. 7.0.1 has been released 2023-12-01.
* The library is part of the Symfony-Framework, a widely used framework which is
  backed by an active community and various companies.

Links:
* Packagist: https://packagist.org/packages/symfony/console
* GitHub: https://github.com/symfony/console
* Documentation: https://symfony.com/doc/current/console.html

klees committed:
Composer: Add seld/jsonlint as dependency

This PR adds `seld/jsonlint` as composer dependency.

Usage:
* This allows the setup to give more informative error messages regarding the
  json that is used for config.

Wrapped By:
* Not applicable, functionality is only used internally in the setup and not
  provided to other ILIAS components.

Reasoning:
* The PHP json parser does not give information where a given json is faulty.
  This degrades user experience since we can only say "your json is wrong" but
  not where the actual error is located.

Maintenance:
* The library receiced its last update in May '23. It seems to be feature
  complete, though.
* The maintainer Jordi Boggiano works on packagist and composer and is a
  well known person in the PHP community.
* The functionality provided by this library is not super important but more
  for quality of life. We could easily decide to ditch the library if any
  problems arise.

Links:
* Packagist: https://packagist.org/packages/seld/jsonlint
* GitHub: https://github.com/Seldaek/jsonlint

klees committed:
Composer: Add ifsnop/mysqldump-php as dependency

This PR adds `ifsnop/mysqldump-php` as composer dependency.

Usage:
* Provides engine to dump database in standard sql for an export.

Wrapped By:
* Not applicable, functionality is only used internally in database service and
  not provided to other ILIAS components.

Reasoning:
* One could use standard mysqldump-tool via `exec` instead, but this would introduce
  an exec call that would be marked during white box sec-analysis.
* The library offers facilities for filtering data, which could become an extra
  benefit sometime, e.g. for creating distributable dumps.

Maintenance:
* The last release of the library was in April '23. There are 45 contributors
  although most contributions are from two people. There is no obvious backing from
  some company or organisation. The code is based on some script that is a lot older.
* This library doesn't seem to be the most reliable on, but since we could easily
  fall back to standard mysql dumper, the risk seems tolerable.

Links:
* Packagist: https://packagist.org/packages/ifsnop/mysqldump-php
* GitHub: https://github.com/ifsnop/mysqldump-php

klees committed:
Composer: Add mustache/mustache as dependency

This PR adds `mustache/mustache` as composer dependency.

Usage:
* Provides templating mechanism for mails, to replace the naive search and replace
  we have used until now.

Wrapped By:
* Not applicable, functionality is only used internally in database service and not
  provided to other ILIAS components.

Reasoning:
* Mustache is in a good place between expressiveness and ease of use. User testing
  has shown that typical administrative users can indeed understand mustache easy
  enough to leverage its power to create mail templates.
* Mustache is a well standardized templating language that has implementations in
  many languages. The specs have been stable for a long time.

Maintenance:
* The library is widely used by many PHP projects. It has 41 contributors, but the
  most contributions have been made by one person. There is no visible backing from
  any organisation. There haven't been any releases for over a year.
* The library doesn't seem to be in a super good place risk wise, but the missing
  releases could also be just a sign for a stable library for a stable spec. Since
  Mustache is a standard and widely use spec we can expect other PHP implementations
  even if this concrete implementation might become unmaintained someday. So the
  risk seems tolerable.

Links:
* Packagist: https://packagist.org/packages/mustache/mustache
* GitHub: https://github.com/bobthecow/mustache.php
* Documentation: https://mustache.github.io/

klees committed:
Composer: Add phpunit/phpunit as dependency

This PR adds `phpunit/phpunit` as composer dependency.

Usage:
* We use it to run unit tests.

Wrapped By:
* Not applicable, provides the binary and classes for test management.

Reasoning:
* Its the standard PHP testing framework.

Maintenance:
* Won't go away any time soon.

Links:
* Documentation: https://phpunit.de

klees committed:
Composer: Add friendsofphp/php-cs-fixer as dependency

This PR adds `friendsofphp/php-cs-fixer ` as composer dependency.

Usage:
* Used in CI and dev processes to check and correct coding style.

Wrapped By:
* Not applicable, provides the binary and classes for cs-fixing.

Reasoning:
* We need some cli-tool to check and correct coding style.

Maintenance:
* The library is under active maintenance.
* It would be non-critical (although quite sad...) to lose this lib.

Links:
* Packagist: https://packagist.org/packages/friendsofphp/php-cs-fixer
* GitHub: github.com/PHP-CS-Fixer/PHP-CS-Fixer

klees committed:
NPM: Add @yaireo/tagify as dependency

This PR adds `@yaireo/tagify` as npm dependency.

Usage:
* Provides functionality for the Tag Input Field of the UI framework.

Wrapped By:
* Tag Input Field

Reasoning:
* A tag input is a powerful tool for certain input scenarios. Building one from
  scratch would be possible but not necessarily efficient.

Maintenance:
* The library has 58 contributors and got its last release in August '23.
* It seems to be feature complete.

Links:
* GitHub: https://github.com/yairEO/tagify

alex40724 committed:
getid3 lib

alex40724 committed:
geshi lib

alex40724 committed:
tinymce lib

Usage:

WYSIWYG text editing.

Wrapped By:

components/ILIAS/COPage (internally)

Reasoning:

Main use is the text element in the page editor. Tiny provides a lot of browser independent dom and editing handling we are relying on. I did not test version 6 yet. The page editor does not need the current fork additions.

Note: This is not about the legacy form integration and its configuration in the ILIAS administration, which is abandoned more or less (hopefully completely in ILIAS 10).

Maintenance:

Well maintained, commercial background, still widely used and integrated.

Links:

Packagist: https://www.npmjs.com/package/tinymce
GitHub: https://github.com/tinymce/tinymce

nhaagen committed:
Composer: add symfony/yaml as dependency

chfsx committed:
NPM: Add `chai` as dependency

chfsx committed:
NPM: Add `dropzone` as dependency

chfsx committed:
NPM: Add `eslint-config-airbnb-base` as dependency

chfsx committed:
NPM: Add `eslint-plugin-import` as dependency

chfsx committed:
NPM: Add `eslint` as dependency

chfsx committed:
Composer: Add `league/commonmark` as dependency

chfsx committed:
NPM: Add `@rollup/plugin-node-resolve` as dependency

chfsx committed:
NPM: Add `@rollup/plugin-terser` as dependency

chfsx committed:
NPM: Add `rollup` as dependency

tfamula committed:
NPM: Add chart.js as dependency

pascalseeland committed:
Composer: Add `apereo/phpcas` as dependency

This PR adds `apereo/phpcas` as composer dependency.

Usage:
* `components/ILIAS/CAS`

Reasoning:
* ILIAS uses this library to enable login via CAS. This lib is provided
  by the same foundation that is also maintaining CAS

Maintenance:
* Not to many commits in the last years, but there are automatic tests
  for php up to 8.2.
* Security issues are always fixed in a timely manner followed by new releases.

Links:
* Packagist: https://packagist.org/packages/apereo/phpcas
* GitHub: https://github.com/apereo/phpCAS
* Documentation: https://apereo.github.io/phpCAS/

mjansenDatabay committed:
NPM/Chatroom: Add `socket.io` and `socket.io-client`

This PR adds `socket.io` and `socket.io-client` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom`
* `components/ILIAS/OnScreenChat`

Wrapped By:
* il.Chat (see: components/ILIAS/OnScreenChat/js/chat.js)
* ServerConnector (internal scope of $.chat, see: components/ILIAS/Chatroom/js/iliaschat.jquery.js)

Reasoning:
* `Socket.IO` is a library that enables low-latency, bidirectional and event-based communication between a client and a server. It is used in the `Node.js` based ILIAS chat server, which relies on the provided `WebSocket` low-level transport mechanisms. `Socket.IO` offers HTTP long-polling as a fallback option, which is useful in environments that don't support WebSockets.

Maintenance:
* `Socket.IO` is a well maintained package with major releases every few years and recent activities.

Links:
* NPM: https://www.npmjs.com/package/socket.io / https://www.npmjs.com/package/socket.io-client
* GitHub: https://github.com/socketio/socket.io / https://github.com/socketio/socket.io-client
* Documentation: https://socket.io/docs/

mjansenDatabay committed:
NPM/Chatroom: Add `winston`

This PR adds `winston` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/Bootstrap/SetupEnvironment.js` (instance creation and configuration)
* `components/ILIAS/Chatroom/chat/*` The `logger` instance is used in many JS files.

This PR adds `winston` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/Bootstrap/SetupEnvironment.js` (instance creation and configuration)
* `components/ILIAS/Chatroom/chat/*` The `logger` instance is used in many JS files.

Wrapped By:
* Not applicable

Reasoning:
* `winston` is a logging library for the `Node.js` based ILIAS chat server, similar to `Monolog`, which we use in ILIAS for server-side logging. It supports multiple transport channels and logging levels (compliant to `RFC5424`, see: https://datatracker.ietf.org/doc/html/rfc5424).

Maintenance:
* `winston` is a well maintained package with major releases every few years. T

Links:
* NPM: https://www.npmjs.com/package/winston
* GitHub: https://github.com/winstonjs/winston
* Documentation: https://github.com/winstonjs/winston

mjansenDatabay committed:
NPM/Chatroom: Add `node-schedule`

This PR adds `node-schedule` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/Bootstrap/SetupClearMessagesProcess.js`
* `components/ILIAS/Chatroom/chat/Bootstrap/UserSettingsProcess.js`

Wrapped By:
* Not applicable

Reasoning:
* `node-schedule` is a flexible cron-like (and not-cron-like) job scheduler for `Node.js`. It allows us to schedule jobs (arbitrary functions) for execution at specific dates/times, with optional recurrence rules. It only uses a single timer at any given time (rather than reevaluating upcoming jobs every second/minute). It is used to cleanup messages (and correspoding/related data) according to the configured ILIAS deletion policy. Furthermore it is used to pull user preferences from the ILIAS backend.

Maintenance:
* `node-schedule` is a well maintained package with a lot of contributions. However, there were only few commits in the last months.

Links:
* NPM: https://www.npmjs.com/package/node-schedule
* GitHub: https://github.com/node-schedule/node-schedule
* Documentation: https://github.com/node-schedule/node-schedule#readme

mjansenDatabay committed:
NPM/Chatroom: Add `async`

This PR adds `async` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/*`

Wrapped By:
* Not applicable

Reasoning:
* `async` is a utility module which provides straight-forward, powerful functions for working with asynchronous JS.

Maintenance:
* `async` is a well maintained package with a lot of contributions and recent activities.

Links:
* NPM: https://www.npmjs.com/package/async
* GitHub: https://github.com/caolan/async
* Documentation: https://caolan.github.io/async/v3/

mjansenDatabay committed:
NPM/Chatroom: Add `mysql`

This PR adds `mysql` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/Persistence/Database.js`

Wrapped By:
* `Database` (see: components/ILIAS/Chatroom/chat/Persistence/Database.js)

Reasoning:
* `mysql` is a mysql driver for `Node.js`. It is used to communicate with a mysql (or mariadb) server.

Maintenance:
* `mysql` has a lot of contributions. However, the last activity is from 2022. This could be a potential risk for new releases of `Node.js` in the future or potential security issues. If we could not rely on this library anymore, we will have to check if there are good alternatives, or we will have to switch to an API based approach on the ILIAS side instead of using a direct database connection.

Links:
* NPM: https://www.npmjs.com/package/mysql
* GitHub: https://github.com/mysqljs/mysql
* Documentation: https://github.com/mysqljs/mysql#readme

mjansenDatabay committed:
NPM/Chatroom: Add `node-mysql`

This PR adds `node-mysql` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/Persistence/Database.js`

Wrapped By:
* `Database` (see: components/ILIAS/Chatroom/chat/Persistence/Database.js)

Reasoning:
* `node-mysql` is an enhancement/extension to the `mysql` library to make it a bit easier to use. It is based on the existing functionalities of (npm) `mysql`.

Maintenance:
* `node-mysql` has contributions provided by only one maintainer. The last activity is from 2015. This could be a potential risk for new releases of `Node.js` in the future or potential security issues. If we could not rely on this library anymore, we will have to check if there are good alternatives, or we just have to remove it from out dependencies and use the original interfaces provided by the `mysql` library.

Links:
* NPM: https://www.npmjs.com/package/node-mysql
* GitHub: https://github.com/redblaze/node-mysql
* Documentation: https://github.com/redblaze/node-mysql/blob/master/README.md

mjansenDatabay committed:
NPM/Chatroom: Add `uuid`

This PR adds `uuid` as NPM dependency for the chatroom

Usage:
* `components/ILIAS/Chatroom/chat/Persistence/Conversation.js`
* `components/ILIAS/Chatroom/chat/Persistence/ConversationAddUser.js`
* `components/ILIAS/Chatroom/chat/Persistence/ConversationMessage.js`

Wrapped By:
* Not applicable

Reasoning:
* `uuid` will be used to retrieve `UUID` values for the `Node.js` based ILIAS chat server, similar to the `ramsey/uuid` library we use in PHP for the ILIAS backend. It will replace the `node-uui` library (which is still used for that chat server).

Maintenance:
* `uuid`  is a well maintained package with a lot of contributions. It is an active project, the latest changes are from October.

Links:
* NPM: https://www.npmjs.com/package/uuid
* GitHub: https://github.com/uuidjs/uuid
* Documentation: https://github.com/redblaze/node-mysql/blob/master/README.md

chfsx committed:
Composer: Add `guzzlehttp/psr7` as dependency

mjansenDatabay committed:
NPM: Add `jQuery`

This PR adds `jQuery` as NPM dependency.

Usage:
* In many components, searching for `$(` results in 2414 occurrences

Wrapped By:
* Not applicable

Reasoning:
* `jQuery` has been introduced many years ago as a
DOM abstraction and utility (XMLHttpRequest, Event Handling etc.) library.
Nowadays, many of this functions could be replaced
by the "Vanilla JS" API. Nevertheless, I doubt that we could tackle the migration
in a timely manner for ILIAS 10. It is unlikely to remove this dependency for
all components outside of a big project with according funding and
organisational resources.

Maintenance:
* `jQuery` is actively maintained and has an existing
security policy (https://github.com/jquery/jquery/blob/e06ff08849057cd099365bf43598c8952fe9956d/SECURITY.md).

Links:
* NPM: https://www.npmjs.com/package/jquery
* GitHub: https://github.com/jquery/jquery
* Documentation: https://api.jquery.com/

mjansenDatabay committed:
NPM: Add `moment`

This PR adds `moment` as NPM dependency.

Usage:
* Calendar: ./components/ILIAS/Calendar/classes/class.ilCalendarUtil.php
* Legacy Forms: Implicitly by `ilCalendarUtil` in the `ilDateTimeInputGUI` field type
* OnScreenChat: ./components/ILIAS/OnScreenChat/js/moment.js
  * It is mainly (but not only) used to display relative dates and
especially times (locale aware) in the conversation windows.
These date/time presentations are dynamically updated
(without expensive server round-trips) every 60 seconds:

```js
			setInterval(() => {
				$('[data-livestamp]').each(() => {
					let $this = $(this);
					$this.html(dateTimeFormatter.fromNowToTime($this.data("livestamp")));
				});
				$('[data-message-time]').each(() => {
					let $this = $(this);
					$this.attr("title", dateTimeFormatter.format($this.data("message-time", "LT")));
				});
			}, 60000);
```

Wrapped By:
* OnScreenChat: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/components/ILIAS/OnScreenChat/js/moment.js

Reasoning:
* `moment` is a JavaScript library to parse, validate, manipulate and
display dates and times in JavaScript.
It has been initially introduced by the `Calendar` component and
is still used there. I guess it has been added as dependency because
at that time JavaScript did not support any locale aware date/time formatting.
As a project we could agree upon a migration to the native
JavaScript `Intl.DateTimeFormat` and `Intl.RelativeTimeFormat` packages
for client-side date.

Maintenance:
* `moment` is actively maintained.

Links:
* NPM: https://www.npmjs.com/package/moment
* GitHub: https://github.com/moment/moment
* Documentation: https://momentjs.com/docs/

mjansenDatabay committed:
NPM: Add `linkifyjs`

This PR adds `linkifyjs` as NPM dependency.

Usage:
* CoPage: Used in paragraphs
* LearningModule: Used for the export
* OnScreenChat: Used to make links in chat messages clickable.
  * See: components/ILIAS/OnScreenChat/js/onscreenchat.js

Wrapped By:
* ./components/ILIAS/Link/js/ilExtLink.js

Reasoning:
* `linkifyjs` is used by the `OnScreenChat` to transform
URL-like strings into clickable HTML link elements. Without this library
all kind of URL-like strings (see: https://linkify.js.org/) are not
transformed anymore and our users will have to copy the strings
manually into the browser address input. I doubt we can support
this feature with a few regular expressions maintained by the ILIAS
community.

Maintenance:
* `linkifyjs` is actively maintained, although it is feature-complete. In the
last months a few bug fixes and improvements have been committed.

Links:
* NPM: https://www.npmjs.com/package/linkifyjs
* GitHub: https://github.com/Hypercontext/linkifyjs
* Documentation: https://linkify.js.org/docs/

thibsy committed:
[FEATURE] Dependencies: add jest as npm package for JavaScript unit-testing.

thibsy committed:
[FEATURE] Dependencies: add @uppy/core as npm package for client-side file uploads.

thibsy committed:
[FEATURE] Dependencies: add @uppy/tus as npm package for chunked file uploads.

chfsx committed:
Composer: Add league/flysystem as dependency

mjansenDatabay commented on components/ILIAS/Chatroom/chat/package.json:
```diff
@@ -3,10 +3,9 @@
   "version": "2.0.0",
   "dependencies": {
-    "async": "^3.2",
+    "async": "3.2",
```
As decided by the JF, we should use v. `3.x` for `winston` (a few lines below this change here), see: https://github.com/ILIAS-eLearning/ILIAS/pull/6769#issuecomment-1850061501.

We already tested this successfully (based on 9.x and an upgraded library and chat component) with v. `^3.13.0`

klees committed:
Composer: add phpunit/phpunit as dependency

This PR adds `phpunit/phpunit` as composer dependency.

Usage:
* We use it to run unit tests.
* ILIAS/Component uses its mocking framework to analyze component definitions
  during build phase.

Wrapped By:
* Not applicable, provides the binary and classes for test management.

Reasoning:
* Its the standard PHP testing framework.

Maintenance:
* Won't go away any time soon.

Links:
* Documentation: https://phpunit.de

---

This initially felt harmless, but an update to PHPUnit 10 basically destroys
our previous approach to the bundling of testcases in testsuites. Currently we
are using PHPUnits TestSuite class in various places, most notably in our global
ILIASSuite, to bundle our cases. That global suite (and some local suites) are
then added to a phpunit.xml to make them available to PHPUnit.

PHPUnit, though, seems to think of the TestSuite class as an internal class that
is used to process the organisation of tests via directories or XML. When using
PHPUnit 10 it complains that we directly add a TestSuite instance via the XML.

Hence, this uses the new order introduced by the Component Revision to allow
PHPUnit to discover tests based on an expression for directories. Individual
components could well add additional phpunit.xmls themselves to implement their
own testssuites. We could also use `--filter` to select certain tests.

klees committed:
PHPUnit: make testcase PHPUnit 10 compatible

* make dataProviders static
* skip testcases with mysterious problems
* fix varios smallish problems
* sometimes tests for T&A fail due to some ordering problem.
  1718723968 is an example seed that shows the behaviour.

chfsx committed:
Composer: Add `dflydev/fig-cookies` as dependency

lukastocker commented:
Hi @kergomard ,

I did some required adjustments.

**Newly included into composer:** 

- league/flysystem
- sabre/dav
- dflydev/fig-cookies
- symfony/console
- simplesamlphp

**Changed:**

- winston to ^3.x (NPM/Chatroom)

The issue regarding symfony/console and simplesamlphp was resolved, as you can see in the comments of following PR:
https://github.com/ILIAS-eLearning/ILIAS/pull/6653

lukastocker committed:
Composer: adjust symfony/yaml and symfony/console package version...

... as symfony/console v7 would not work with any simplesamlphp package
right now, we decided to downgrade symfony/console to v6.4.

Symfony/yaml had to be adjusted accordingly.

The discussion can be read here: https://github.com/ILIAS-eLearning/ILIAS/pull/6653

klees committed:
Component: fix Dependency Reader for PHPUnit 10

klees committed:
Filesystem: StreamInterface compat

smeyer-ilias committed:
Jour Fixe monolog dependency

klees committed:
DEP: clean extras in composer.json

This information is contained in separate commits. Also, we will discuss the dependencies
every year again, so a dates won't be of much relevance anymore.

klees committed:
DEP: uncomment missing npm-dependencies

klees committed:
DEPS: add test-folders to classmap to run tests without explicit 'require'

klees committed:
UI: compat with PHPUnit 10

klees committed:
DEPS: deactivate tests that require missing deps

* mostly these tests depend on Mockery
* some depend on LibRIS or PHPSpreadsheet
* to try them out, remove the "TBD" in the name and run it

klees committed:
Setup: PHPUnit 10 compat

klees committed:
UI: psr7 compat in tests

klees committed:
DEPS: disable js-tests

Migration to jest is pending.

kergomard commented:
Thank you very much @lukastocker and @klees for all the effort you put into this!

There is still a long list of missing js-dependencies, but:
- `npm clean-install` + `composer install` + `php cli/setup.php update -y` run through and lead to a system one can log into and that is generally usable.
- `scripts/PHPUnit/run_tests.sh` can be run. It leads to a lot of deprecation warnings, but it runs.
- `npm test` cannot be run, as we need to move from mocha to jest, but the tests are currently commented out.

I will thus merge this now and we go on from there.

Thanks again and best,
@kergomard in the name of the Technical Board

klees commented:
Dear colleagues,

this PR replaces the dependencies of ILIAS 9 by the dependencies that have been approved by the JF. If we'd missed anything we are sorry, just a lot of changes to incorporate here. Please add missing stuff via PR. There are some open ends that need to be tackled by individual contributors:

* [This commit](https://github.com/ILIAS-eLearning/ILIAS/pull/7141/commits/7aac5c91c665a77601b5fa2fa36a262a73e39b5a) uncomments files that have been pulled from some node-module that hasn't been requested until now. I will take this list to the JF to make it public there, but everybody might want to have a look into this already and tackle the problem for their component. There will be stuff that currently does not work as expected because of missing npm-packages.
* [This commit](https://github.com/ILIAS-eLearning/ILIAS/pull/7141/commits/f07026261d5cb65d2d1ef5143f173bf2e5e53077) introduces many changes to make tests PHPUnit 10 compatible as discussed [here](https://github.com/ILIAS-eLearning/ILIAS/pull/7427). It feels as if we have unearthed various tests that haven't been run before. We tried to fix them, but deliberately added `markTestSkipped` if we felt that there was more than a compatibility problem in some failing test. Please everyone, look into skipped tests in your component and decide what to do with them.
* [This commit here](https://github.com/ILIAS-eLearning/ILIAS/pull/7141/commits/f32923df22eede698fa47b09522e64832460abe9) contains tests that errored or failed due to missing PHP dependencies. This mostly concerns the test-library `Mockery`. Since this makes test fail very early I renamed files so PHPUnit does not pick these tests up anymore. Please look into the according tests. Remove the `TBD` from the filename to make PHPUnit run the test and see the error. From my POV we should try to get rid of `Mockery` completely.
* [This commit](https://github.com/ILIAS-eLearning/ILIAS/pull/7141/commits/3aa813091b14da78633a190037b4cc2f674d003d) deactivates js-tests in our CI pipeline. We decided to [move tests to jest](https://github.com/ILIAS-eLearning/ILIAS/pull/6945) and I currently try to figure out how we would tackle that problem.
* As @kergomard points out, the PHPUnit tests display deprecation warnings. Please look into these, as these will bite us sometimes later.
* There is some hidden dependency in tests of the T&A that makes the tests fail for some execution orders. 1718723968 is a seed that exhibits that behaviour. If you run tests and a lot of tests fail in the T&A: just run them again. You should hit some working seed soon.

That's it for the moment. I'll bring this up on the JF again, but I definitly propose to merge this now. We can install and log into the system. There will be open issues due to missing javascript-libs, but if we do not move on with the freshly approved dependencies we will run into problems very late in our dev-process.

Kind regards!
